[CDAP-18322] Mount secret in sidecar ArtifactLocalizer but not in preview runner and task worker

wyzhang opened the pull request:
Why:
cdap-secret shouldn't be exposed to preview runner and task worker containers
which run user provided and untrusted code. But it can be mounted in sidecar
ArtifactLocalizer container which only run system/trusted code. Mounting
the secret there allows requests originated by ArtifactLocalizer to be authenticated
as system code to perform system-level operations (e.g. fetching artifacts to
prewarm local caches for task worker containers)

masoud-io commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/k8s/MasterEnvironmentMain.java:
```diff
@@ -138,4 +149,35 @@ public static void doMain(String[] args) throws Exception {
+  private static InternalAuthenticator getInternalAuthenticator(CConfiguration cConf) {
+    File sConfFile = new File(cConf.get(Constants.Twill.Security.MASTER_SECRET_DISK_PATH));
+    Injector injector;
+    if (sConfFile.exists()) {
```
What if MASTER_SECRET_DISK_PATH has been set, but the path does not exist? Shouldn't we throw an exception instead of executing the else block?

wyzhang replied:
The condition below is based on if path exists or not, not based on if the config is set.

masoud-io replied:
Correct. But if one makes a mistake in setting the path?

masoud-io commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -282,7 +282,7 @@
   <property>
     <name>twill.security.worker.mount.secret</name>
-    <value>true</value>
+    <value>false</value>
```
Given that we'll have secret mount in sidecar container, is there any scenario in which we want secret mount in worker runnable?

wyzhang replied:
No. Secret shouldn't be exposed to the environment where untrusted code runs

masoud-io commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java:
```diff
@@ -919,17 +919,13 @@ private V1PodSpec createPodSpec(Location runtimeConfigLocation, Map<String, Runt
-      }
-    }
+    // Mount all volumes including cdap-secret for init container as it runs system/trusted code.
+    List<V1VolumeMount> initContainerVolumeMounts = new ArrayList<>(volumeMounts);
```
nit: I assume we don't need secrets for downloading artifacts from GCS in the init container. Right?

wyzhang replied:
It has already been changed to download via appfab

masoud-io commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -461,14 +463,14 @@ private PodInfo createPodInfo(Map<String, String> conf) throws IOException, ApiE
       .findFirst()
       .orElse(pod.getSpec().getContainers().get(0));
 
+    Set<String> configs = ImmutableSet.of(CDAP_CONFIG_NAME, HADOOP_CONFIG_NAME, SECURITY_CONFIG_NAME);
```
nit: why not to reuse CONFIG_NAMES and add CDAP_CONFIG_NAME, HADOOP_CONFIG_NAME and SECURITY_CONFIG_NAME to it.

wyzhang replied:
Done

dli357 commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DistributedPreviewManager.java:
```diff
@@ -151,9 +151,15 @@ public void run() {
+          CConfiguration cConfCopy = CConfiguration.copy(cConf);
           Path cConfPath = runDir.resolve("cConf.xml");
+          if (!cConf.getBoolean(Constants.Twill.Security.WORKER_MOUNT_SECRET)) {
+             // Unset the internal certificate path since certificate is stored cdap-security which
```
Can we open a JIRA and track this as a TODO, as this will no longer work properly when SSL verification is enabled?

wyzhang replied:
Done

dli357 commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/k8s/MasterEnvironmentMain.java:
```diff
@@ -113,7 +123,8 @@ public static void doMain(String[] args) throws Exception {
         RemoteClientFactory remoteClientFactory = new RemoteClientFactory(
           masterEnv.getDiscoveryServiceClientSupplier().get(),
-          new DefaultInternalAuthenticator(new WorkerAuthenticationContext()));
+          getInternalAuthenticator(cConf));
```
Instead of running the sidecar and primary containers on the same main class, maybe it would be cleaner and more maintainable to separate them into two main classes? This can be done in a future PR as well.

wyzhang replied:
Ack

dli357 commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -111,9 +111,10 @@
   static final String NAMESPACE_CPU_LIMIT_PROPERTY = "k8s.namespace.cpu.limits";
   static final String NAMESPACE_MEMORY_LIMIT_PROPERTY = "k8s.namespace.memory.limits";
 
+  public static final String SECURITY_CONFIG_NAME = "cdap-security";
```
I think we may run into problems if we hardcode the `cdap-security` volume mount name as it is actually specified in the CDAP CR, so we may want to make this configurable.

wyzhang replied:
Ack

dli357 commented on cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:
```diff
@@ -111,9 +111,10 @@
   // needed to propagate to deployments created via the KubeTwillRunnerService
-  private static final Set<String> CONFIG_NAMES = ImmutableSet.of("cdap-conf", "hadoop-conf");
+  private static final Set<String> CONFIG_NAMES = ImmutableSet.of("cdap-conf", "hadoop-conf", "cdap-security");
   private static final Set<String> CUSTOM_VOLUME_PREFIX = ImmutableSet.of("cdap-cm-vol-", "cdap-se-vol-");
```
Can be done in a future PR, but can we add a volume prefix for security-sensitive custom volumes which should be mounted to sidecar/privileged containers but not containers which run user code? That would make it a bit more flexible for us in the future.

wyzhang replied:
Ack

dli357 reviewed (approved):
Minor comments, otherwise LGTM. Thanks!

wyzhang committed:
[CDAP-18322] Mount secret in sidecar ArtifactLocalizer but not in preview runner and task worker

Why:
cdap-secret shouldn't be exposed to preview runner and task worker containers
which run user provided and untrusted code. But it can be mounted in sidecar
ArtifactLocalizer container which only run system/trusted code. Mounting
the secret there allows requests originated by ArtifactLocalizer to be authenticated
as system code to perform system-level operations (e.g. fetching artifacts to
prewarm local caches for task worker containers)
